### ISSUE TITLE
[2018 ICLR] Enhancing The Reliability of Out-of-distribution Image Detection in Neural Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#60]
ODIN 논문

Temperature scaling과 input perturbation을 통해 in-distribution / out-of-distribution을 잘 구분할 수 있음을 보여줌

## 1. Problem Statement

Pre-trained neural network를 통해 in-distribution / out-of-distribution images인지 구분하는 성능을 올리는데만 집중

In-distribution correct / incorrect는 in-distribution으로 detect되면 original image 넣어서 classify하면 되는 쉬운 문제이기에 고려 X

## 2. ODIN : Out-of-distribution Detector

Detector의 핵심 2가지 파트는 다음과 같음
1. Temperature scaling
2. Input preprocessing

### 2.1. Temperature scaling
![image](https://user-images.githubusercontent.com/49643709/161222652-1993bc78-d106-4b4a-a815-1092e0e28b24.png)

Temperature scaling을 통해 in-distribution / out-of-distribution images의 softmax scores를 더 separable하게 만들어줌

### 2.2. Input preprocessing
![image](https://user-images.githubusercontent.com/49643709/161222875-1f8e160c-0346-4783-9501-3dcb1cf9bdfd.png)

Adversarial attack에서는 input에 perturbation을 더해 softmax score를 감소시킨 것에 intuition을 가져옴

Adversarial attack과 반대로 input에 perturbation을 더해 in-distribution / out-of-distribution images의 softmax scores를 더 separable하게 만들어줌

### 2.3. Out-of-distribution detector
![image](https://user-images.githubusercontent.com/49643709/161223607-ea1a9bcd-fa5b-4e50-8c09-938ffb95ce28.png)

1. Calculate preprocessed image
2. Feed the preprocessed image into the neural network
3. Calculate calibrated softmax score
4. Compare the score to the threshold and classify if an image is in-distribution / out-of-distribution

모든 hyperparameter (softmax temperature, perturbation magnitude, threshold)는 TPR이 95%일 때 FPR이 낮게 나오는 값을 선택

#### 3.1.1. Accuracy of pre-trained networks
<img src='https://user-images.githubusercontent.com/49643709/161225800-a23eaef2-2071-44ad-9251-851c8b102a7b.png' width=60% height=60%>

#### 3.1.2. Out-of-distribution datasets

- TinyImageNet (crop, resize) : randomly crop 32 x 32 image patch / downsample to 32 x 32 image
- LSUN (crop, resize) : randomly crop 32 x 32 image patch / downsample to 32 x 32 image
- Gaussian Noise : synthetic noise where each pixel is sampled from i.i.d. Gaussian distribution with mean 0.5 and unit variance (each pixel is clipped into range [0, 1])
-  Uniform Noise : synthetic noise where each pixel is sampled from i.i.d. uniform distribution (each pixel is clipped into range [0, 1])

Hyperparameter는 OOD test dataset과 independent한 iSUN dataset을 이용하여 tuning
(iSUN images are downsampled to 32 x 32)

#### 3.1.3. Evaluation metrics

- FPR at 95% TPR : TPR이 95%일때 의 FPR
- Detection Error : TPR이 95%일 때의  misclassification probability (P_e = (1 - TPR) / 2 + FPR / 2)
- AUROC
- AUPR

### 3.2. Comparison with baseline
<img src='https://user-images.githubusercontent.com/49643709/161230824-c0ea478a-dec6-46b5-a212-199572ef3eee.png' width=100% height=100%>
<img src='https://user-images.githubusercontent.com/49643709/161230577-62715957-688e-4899-9bdd-3829ca69ab55.png' width=40% height=40%>

다양한 dataset에서 모두 성능이 향상됨

### 3.3. Parameter transferability
![image](https://user-images.githubusercontent.com/49643709/161231085-36d20b39-5a3c-4e77-ac65-5f6069cc281a.png)

Hyperparameter가 어떤 validation dataset에서 찾았는지와 상관없이 generalize 잘됨

### 3.4. Data distributional distance vs detection performance
![image](https://user-images.githubusercontent.com/49643709/161258682-cdd86c35-ac20-4d27-a161-85d92b6f44c4.png)

In-distribution dataset과 out-of-distribution dataset간의 statistical distance를 측정하기 위해 maximum mean discrepancy (MMD) with Gaussian RBF kernel, energy distance를 사용

#### 3.4.1. MMD distance
![image](https://user-images.githubusercontent.com/49643709/161231840-c64b5fe7-7f99-49dd-89e9-869caecb9545.png)
![image](https://user-images.githubusercontent.com/49643709/161231907-d3f263bd-e012-4450-950f-51e7635dd105.png)

1. MMD distance between CIFAR-100 and out-of-distribution datasets
    - Cropped images는 local image context를 가져 CIFAR-100 이미지와 유사하지 않기에 CIFAR-100과 cropped dataset 사이의 MMD distance가 크다고 분석할 수 있음
    - 그에 반해 resized images는 global patterns을 가지며 CIFAR-100 이미지와 유사하기에 CIFAR-100과 cropped dataset 사이의 MMD distance가 작다고 분석할 수 있음
2. MMD distance tends to negatively correlated with detection performance
    - MMD distance가 작다는 말은 in-distribution과 out-distribution이 서로 유사하다는 의미이기에 detect하기가 어려움

#### 3.4.2. Energy distance
![image](https://user-images.githubusercontent.com/49643709/161258420-2e21c499-dce3-409c-aa97-6f45d0b0782b.png)

#### 4.1.1. Taylor expansion
![image](https://user-images.githubusercontent.com/49643709/161248210-f03742e1-d24d-448b-97bf-fddd0b678367.png)
![image](https://user-images.githubusercontent.com/49643709/161248286-ddcc8618-0c1c-4126-ab2c-c1eb39a96572.png)
![image](https://user-images.githubusercontent.com/49643709/161258000-8b96399f-8925-4061-872c-acc57626227e.png)

Taylor expansion을 통해 softmax score을 U1과 U2의 식으로 근사화할 수 있음

근사화한 식을 통해 softmax score는 U1에 비례하며 -U2에 비례함을 알 수 있음

U1은 largest unnormalized output이 다른 remaining output에 비해 벗어난 정도를 측정함
(U1 measures the extent to which the largest unnormalized output of the neural network deviates from the remaining outputs)

U2는 remaining smaller output이 서로 얼마나 퍼져있는지를 측정함
(U2 measures the extent to which the remaining smaller outputs deviate from each other)

#### 4.1.2. Behavior of U1 and U2
![image](https://user-images.githubusercontent.com/49643709/161251942-f9a7e3b0-5fea-459e-ac77-2f7f2df3ca94.png)
![image](https://user-images.githubusercontent.com/49643709/161257480-ccd1f60a-4585-41ff-870c-e6b74e403b52.png)

Temperature가 1일 때 U1 값은 in-distribution, out-of-distribution 서로 비슷함

하지만 U2 값은 in-distribution일 때가 out-of-distribution보다 값이 큼

이가 의미하는 바는 in-distribution images의 경우, remaining output이 out-of-distribution보다 더 separated되어있다는 뜻임

In-distribution dataset 내에서 몇몇 class들이 다른 class와는 확연히 다르게 common feature를 share하기 때문에 나타나는 현상임

예를 들면 dog image가 주어졌을 때 label dog는 label cat과 가까우며 label car, truck과는 separated되어있음

#### 4.1.3. The effects of T
![image](https://user-images.githubusercontent.com/49643709/161251178-75b96551-c981-43eb-a0fd-139e41e60fb0.png)

즉 in-distribution의 특성으로 인해 out-of-distribution보다 U2 값이 큼

이는 softmax score을 작게 만들어 in-distribution과 out-of-distribution을 detect하는데 부정적인 영향을 줌

Temperature 값을 크게 만들어 U2의 영향력을 줄여 in-distribution softmax score 값을 크게 만들어 이를 해결

이로인해 detect performance가 좋아짐

#### 4.2.1. Taylor expansion
![image](https://user-images.githubusercontent.com/49643709/161255451-18f99a9a-81f1-4a45-a7b1-58780f01b269.png)

Taylor expansion을 통해 perturbed image x에 대한 log-softmax function을 다음과 같이 쓸 수 있음

#### 4.2.2. The effects of gradient
<img src='https://user-images.githubusercontent.com/49643709/161257676-89ba8e76-f880-4662-bf36-2a8608657637.png' width=100% height=100%>
<img src='https://user-images.githubusercontent.com/49643709/161257882-bf90536c-19b5-49ff-8813-b7765db004d4.png' width=100% height=100%>
<img src='https://user-images.githubusercontent.com/49643709/161256320-f4f928cd-d4b8-49df-82bb-054de97fbc93.png' width=40% height=40%>

In-distribution images는 상대적으로 out-of-distribution images보다 gradient norm이 더 큰 값을 가지는 경향성이 있음을 관측함

즉 in-distribution softmax score이 out-of-distribution softmax score과 비슷하더라도 in-distribution의 gradient norm이 더 큼

Input preprocessing을 통해 in-distribution images와 out-of-distribution images를 더 separable하게 만들 수 있음

#### 4.2.3. The effect of ε
![image](https://user-images.githubusercontent.com/49643709/161255506-34b26a15-7d6b-4450-b010-587af3bae50b.png)

Image에 충분히 작은 perturbation을 더하는 것은 prediction을 바꾸지 않지만 in-distribution image의 경우 softmax score를 significantly하게 바꿔줌

이로 인해 detect performance가 좋아짐

#### 4.3.1. The effect of threshold
![image](https://user-images.githubusercontent.com/49643709/161258882-4e0890d3-bacb-4748-b775-c4bbae08f7ae.png)

95% TPR에 해당하는 threshold 값을 선택하면 all out-of-distribution datasets에서 충분히 작음 FPR을 가질 수 있음

#### 4.3.2. Difficult-to-classify images and difficult-to-detect images
![image](https://user-images.githubusercontent.com/49643709/161221518-ae9d4538-7912-4cb0-91b8-b40c9639ddf7.png)

Detect하기 어려운 images들은 상대적으로 더 classify하기 어려움
(images that are difficult to detect are more likely to be the images that are difficult to classify)
